src/mem/heap/heap_simple.c

```diff
@@ -125,7 +129,7 @@ void *malloc(size_t size) {
 				mcb->size = size;
 			}
 
-			return ptr + MCB_SZ;
+			return (void *)(((size_t)ptr + MCB_SZ + boundary -1) & ~(boundary-1));
```

Comment:
Are you sure it's correct?

Look, what memory structure we have:
```
( MCB0 | object0 ) ( MCB1 | object1 ) ....
```

where MCB (Memory Control Block) is `struct mem_control_block`.

In more details:

```
( is_available0; size0 | object0 ) ( is_available1; size1 | object1 ) ...
```

Here, size0 is the size of object0, size1 is the size of object1, etc.

Ok, but what happens when you add `boundary` to ptr here?

```
( is_available0; size0 | <shift to boundary> object0 ) ( is_available1; size1 | object1 ) ...
```

So because now the actual size of object0 is _not_ size0 like it was before, but is `<shift to boundary> + size0`

Reply:
I think it's correct because we choose the memory block in which our offset fits and return the aligned address. Everything in this structure remains as before, we have only a few bytes of unused space before the address.

Reply:
Ok, now I understand, you have right, I will fix it.

Reply:
The amount of commits doesn't matter right now, do not worry. They can be squashed into a single one then.

I will take a look right now, this part of the code is not very simple, so we should be careful.

Reply:
Yes, it is correct now.

Reply:
But it seems like `free()` will broken now.

Reply:
Because of that:

```
void free(void *ptr) {
	struct mem_control_block *mcb;

	mcb = (void *)((char *) ptr - MCB_SZ);
...
```

Reply:
Maybe we can add a parameter to mcb "shifted" and then read it when we use free?

Reply:
But how you will get the pointer to mcb only from the ptr passed into free() in this case? The shift is variable in size. We should save some information along with the allocated ptr to obtain mcb then.

Reply:
At now we can do it like this
MCB_SZ shift pointer                  
and return ptr so I want do it like this
shift MCB_SZ pointer
We can read mcb and free aligned bytes. I will test it.

Reply:
For example,

```
mcb <padding> object
```

So, `address of mcb = object - <padding> - MCB_SZ`. But the padding here depends on the shift value you calculated before in mmemalign. So probably you should do something like this:

```
mcb <padding> <shift value> object
```

So, you can get shift value first from ptr, and then get mcb..

Reply:
Yes, or try to do it in the way you proposed. Any solution would be great.

Reply:
Yes, please try your variant first.

Reply:
But in your variant is another potential problem - with _mem_defrag() called from free(). Here we go over all free blocks and concatenate them.

```
shift0 mcb0 ...
```

how we can get the first mcb0?

Reply:
I think, your solution is better and less complicated, I will use it :).

Reply:
I added a fix, could you take a look? Are all tests from test cmd enough?

Thanks.

Reply:
It looks great now!